huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
+                                      'before next exposure.')
+                    max_num_exposures += 1
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _take_flat_fields(self, exptimes, observation, fits_headers=None, dark=False):
+        """
+        Slew to flat field, take exposures and wait for them to complete.
+        Returns a list of camera events for each camera.
+
+        args:
+            exptimes: dict of camera_name: list of exposure times.
+            observation: Flat field Observation object.
+        """
+        image_dir = self.config['directories']['images']
+        imtype = 'dark' if dark else 'flat'
+        if fits_headers is None:
+            fits_headers = self.get_standard_headers(observation=observation)
+
+        # Slew to field
+        self.logger.debug(f'Slewing to {imtype}-field coords: {observation.field}.')
+        self.mount.set_target_coordinates(observation.field)
+        self.mount.slew_to_target()
+        self.status()  # Seems to help with reading coords
+
+        # Loop over cameras...
+        camera_events = {}
+        for cam_name, exptime in exptimes.items():
+            cam = self.cameras[cam_name]
+
+            # Define filename for flat fields
+            filename = (f'{image_dir}/flats/{cam.uid}/{observation.seq_time}'
+                        f'/{imtype}_{observation.current_exp_num:02d}.{cam.file_extension}')
+
+            # Take exposure and get event
+            exptime = exptime.to(u.second).value
+            camera_event = cam.take_observation(observation, fits_headers,
+                                                filename=filename, exptime=exptime)
+            camera_events[cam_name] = {'event': camera_event, 'filename': filename}
+
+        # Block until done exposing on all cameras
+        while not all([info['event'].is_set() for info in camera_events.values()]):
+            self.logger.debug('Waiting for flat-field images...')
+            time.sleep(1)
+
+        return camera_events
+
+    def _take_flat_field_darks(self, exptimes, observation):
+        """Take the dark fileds for each camera.
+
+        args:
+            exptimes: dict of camera_name: list of exposure times.
+            observation: Flat field Observation object.
+        """
+        # Exposure time lists may not be the same length for each camera
+        while True:
+            next_exptimes = {}
+            for cam_name, exptime in exptimes.items():
+                try:
+                    next_exptimes[cam_name] = exptimes[cam_name].pop(0)
+                except IndexError:  # No more exposures left for this camera
+                    continue
+                self._take_flat_fields(next_exptimes, observation, dark=True)
+            # Break if we have finished all the cameras
+            if not next_exptimes.items():
+                break
```

Comment:
I think I spot a few problems here, shouldn't the call the `_take_flat_fields()` be outside the `for` loop, and after the check to see if the darks are all finished?

Hard to make this much more readable, but I came up with this:

```suggestion
        # Exposure time lists may not be the same length for each camera
        for i in itertools.count():
            next_exptimes = {}
            for cam_name, exptime in exptimes.items():
                with suppress(IndexError):
                    next_exptimes[cam_name] = exptimes[cam_name][i]
            # Break if we have finished all the cameras
            if not next_exptimes:
                break
            # Take the next set of darks with the cameras
            self._take_flat_fields(next_exptimes, observation, dark=True)
```

Reply:
Yes, absolutely.